Modularization

wirew0rm opened the pull request:
Efforts to make chartfx usable from jigsaw modularized projects as proposed in #156.

Before merging this the following points will have to be addressed:

## Jigsaw Compatibility

To make chartfx usable for modularized projects, there are some points which have to be considered according to this [blogpost](http://branchandbound.net/blog/java/2017/12/automatic-module-name/).

> Addressing these concerns is a matter of good hygiene. If you encounter one of these issues and you can't address those, don't add the Automatic-Module-Name entry yet. For now, your library is better off on the classpath. It will only raise false expectations if you do add the entry.

So before meging this branch, these issues should probably be adressed.

## internal JDK types

> Make sure your library doesn't use internal types from the JDK (run jdeps --jdk-internals mylibrary.jar to find offending code). JDeps (as bundled with Java 9 and later) will offer publicly supported alternatives for any use of encapsulated JDK APIs. When your library runs from the classpath on Java 9 and later, you can still get away with this. Not so if your library lives on the module path as automatic module.

```
jdeps --jdk-internals chartfx-chart/target/chartfx-chart-11.1.0-SNAPSHOT.jar 
chartfx-chart-11.1.0-SNAPSHOT.jar -> JDK removed internal API
   de.gsi.chart.utils.SimplePerformanceMeter          -> com.sun.javafx.perf.PerformanceTracker             JDK internal API (JDK removed internal API)

jdeps --jdk-internals chartfx-dataset/target/chartfx-dataset-11.1.0-SNAPSHOT.jar 
chartfx-dataset-11.1.0-SNAPSHOT.jar -> jdk.unsupported
   de.gsi.dataset.serializer.spi.FastByteBuffer       -> sun.misc.Unsafe                                    JDK internal API (jdk.unsupported)

JDK Internal API                         Suggested Replacement
----------------                         ---------------------
sun.misc.Unsafe                          See http://openjdk.java.net/jeps/260
```

Unsafe replacement links to [Variable Handles](http://openjdk.java.net/jeps/193).

 - [x] com.sun.javafx.perf.PerformanceTracker (chartfx-chart: SimplePerformanceMeter)
 - [x] sun.misc.Unsafe (chartfx-datasets: FastByteBuffer)

## unnamed packages

> Your library can't have classes in the default (unnamed) package. This is a bad idea regardless, but when your library is used as automatic module, this rule is enforced by the module system.

  - [x] no problems here

## split packages

> Your library can't split packages (two or more JARs defining the types in the same package), nor can it redefine JDK packages (javax.annotation is a notorious example, being defined in the JDK's java.xml.ws.annotation module but also in external libraries).

Are our testcases affecting this? They are in the same package as the implementation but they are not shipped, so it might be ok?

## META-INF/services

> When your library's JAR has a META-INF/services directory to specify service providers, then the specified providers must exist in this JAR (as described in the ModuleFinder JavaDoc)

  - [x] check

ennerf commented:
Regarding the compatibility questions:

**Split Packages**
Test cases don't matter

**Internal JDK Types**
`sun.misc.unsafe` is still available for the time being, so it's probably not necessary to replace all references just to get the automatic module name in.

 `VarHandles` is only applicable to some of the on-heap use cases Unsafe. In case you're doing anything off-heap, you may need to take a look at the new [JEP 370: Foreign-Memory Access API](https://openjdk.java.net/jeps/370).

RalphSteinhagen commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 9
           

Complexity increasing per file
==============================
- chartfx-math/src/test/java/de/gsi/math/filter/fir/FirFilterTests.java  2
- chartfx-math/src/main/java/de/gsi/math/filter/fir/FirFilter.java  16
         

Complexity decreasing per file
==============================
+ chartfx-math/src/main/java/de/gsi/math/ArrayMath.java  -13
+ chartfx-math/src/test/java/de/gsi/math/ArrayMathTests.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=5382087&bid=17646823)

wirew0rm commented:
We decided to target this for the next major release, so rebased against dev-x.2.0.

Also tried to go fully modularized. This removes the need for our custom launcher, but we still need some jvm arguments, mostly due to controlsfx using internal javafx api via reflection.

The serialisers also rely on reflection, so there are still some exports missing.
Also correctly structuring the tests is difficult, as they cannot be in a different module, because they are in the same package but we cannot require the test dependencies from the module-info because they are only in the maven `test`scope. Surefire seems to do some magic here by supplying some additional jvm parameters, but at least in combination with testFX it does not work out of the box.

On the bright side, we can now start some of our samles without any additional configuration or wrappers.

wirew0rm commented:
Projects missing auto-moudule-name in manifest:
```
[WARNING] Can't extract module name from batik-script-1.12.jar: Provider class org.apache.batik.bridge.RhinoInterpreterFactory not in module
[WARNING] Can't extract module name from xalan-2.7.2.jar: Provider class org.apache.bsf.BSFManager not in module
[WARNING] * Required filename-based automodules detected: [JTransforms-3.1.jar, commons-math3-3.6.1.jar, javafxsvg-1.3.0.jar, pngj-2.1.0.jar]. Please don't publish this project to a public artifact repository! *
```
- [JTransforms](https://github.com/wendykierp/JTransforms): no efforts yet
- apache commons-math: fixed in upstream but unreleased apache/commons-math#107
- [javafxsvg](https://github.com/codecentric/javafxsvg): to be removed and replaced by iconli, also depends on internal api
- [pngj](https://github.com/leonbloy/pngj): no efforts yet

RalphSteinhagen commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 7
           

Coverage increased per file
===========================
+ chartfx-chart/src/main/java/de/gsi/chart/ui/ProfilerInfoBox.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=5382087&bid=17646823)

RalphSteinhagen commented on JigsawCompatibility.md:
```diff
@@ -0,0 +1,48 @@
+
+Unsafe replacement links to [Variable Handles](http://openjdk.java.net/jeps/193).
+
+ - [ ] com.sun.javafx.perf.PerformanceTracker (chartfx-chart: SimplePerformanceMeter)
```
Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=5382087)

RalphSteinhagen replied:
Codacy found an issue: [[list-item-bullet-indent] Incorrect indentation before bullet: remove 1 space](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=5382087)

RalphSteinhagen commented on JigsawCompatibility.md:
````diff
@@ -0,0 +1,48 @@
+
+> Make sure your library doesn't use internal types from the JDK (run jdeps --jdk-internals mylibrary.jar to find offending code). JDeps (as bundled with Java 9 and later) will offer publicly supported alternatives for any use of encapsulated JDK APIs. When your library runs from the classpath on Java 9 and later, you can still get away with this. Not so if your library lives on the module path a ...
+
+```
````
Codacy found an issue: [[fenced-code-flag] Missing code-language flag](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=5382087)

RalphSteinhagen commented on JigsawCompatibility.md:
```diff
@@ -0,0 +1,48 @@
+
+> Your library can't have classes in the default (unnamed) package. This is a bad idea regardless, but when your library is used as automatic module, this rule is enforced by the module system.
+
+  - [x] no problems here
```
Codacy found an issue: [[list-item-bullet-indent] Incorrect indentation before bullet: remove 2 spaces](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=5382087)

RalphSteinhagen commented on JigsawCompatibility.md:
```diff
@@ -0,0 +1,48 @@
+
+> When your library's JAR has a META-INF/services directory to specify service providers, then the specified providers must exist in this JAR (as described in the ModuleFinder JavaDoc)
+
+  - [x] check
```
Codacy found an issue: [[list-item-bullet-indent] Incorrect indentation before bullet: remove 2 spaces](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=5382087)

RalphSteinhagen commented on JigsawCompatibility.md:
```diff
@@ -0,0 +1,48 @@
+Unsafe replacement links to [Variable Handles](http://openjdk.java.net/jeps/193).
+
+ - [ ] com.sun.javafx.perf.PerformanceTracker (chartfx-chart: SimplePerformanceMeter)
+ - [ ] sun.misc.Unsafe (chartfx-datasets: FastByteBuffer)
```
Codacy found an issue: [[list-item-bullet-indent] Incorrect indentation before bullet: remove 1 space](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=5382087)

RalphSteinhagen replied:
Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=5382087)
